### ISSUE TITLE
Row fetches can finish before fields query

### ISSUE THREAD

**Issue by timwis** [JarvusInnovations/jarvus-postgrest-apikit#5]
I'm noticing that my app consistently finishes `store.load()` before the `OPTIONS` request in `model/Postgrest.js`. Is there any way to hook to the `OPTIONS` event completing from my view? I was looking for some event that gets fired by `model.addFields()` perhaps, but there don't appear to be any.

**Comment by themightychris** [comment]
This shouldn't be possible, the OPTIONS request is done as part of the
async class-loading cycle, so if you're requiring the model from the store
(or in the case of an inline initialized store requiring the model from the
containing class) the definition of the dependent class should be deferred
until the model is finished loading. Can you link to a test case?

On Sat, Aug 15, 2015 at 11:48 PM, Tim Wisniewski notifications@github.com
wrote:

> I'm noticing that my app consistently finishes store.load() before the
> OPTIONS request in model/Postgrest.js. Is there any way to hook to the
> OPTIONS event completing from my view? I was looking for some event that
> gets fired by model.addFields() perhaps, but there don't appear to be any.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/JarvusInnovations/jarvus-postgrest-apikit/issues/5.

## 

Chris Alfano
Chief Developer
Jarv.us Innovations http://jarv.us

(e) chris@jarv.us (m) 267-918-6166

**Comment by themightychris** [comment]
Have you noticed if this behavior is different between the built and
development versions of the app?

On Sun, Aug 16, 2015 at 8:44 AM, Chris Alfano chris@jarv.us wrote:

> This shouldn't be possible, the OPTIONS request is done as part of the
> async class-loading cycle, so if you're requiring the model from the store
> (or in the case of an inline initialized store requiring the model from the
> containing class) the definition of the dependent class should be deferred
> until the model is finished loading. Can you link to a test case?
> 
> On Sat, Aug 15, 2015 at 11:48 PM, Tim Wisniewski <notifications@github.com
> 
> > wrote:
> > 
> > I'm noticing that my app consistently finishes store.load() before the
> > OPTIONS request in model/Postgrest.js. Is there any way to hook to the
> > OPTIONS event completing from my view? I was looking for some event that
> > gets fired by model.addFields() perhaps, but there don't appear to be
> > any.
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/JarvusInnovations/jarvus-postgrest-apikit/issues/5.
> 
> ## 
> 
> Chris Alfano
> Chief Developer
> Jarv.us Innovations http://jarv.us
> 
> (e) chris@jarv.us (m) 267-918-6166

## 

Chris Alfano
Chief Developer
Jarv.us Innovations http://jarv.us

(e) chris@jarv.us (m) 267-918-6166

**Comment by timwis** [comment]
Yes - in the development version (`sencha app watch`) this doesn't occur and it functions as expected. In the built production version it happens very often. In the screenshot below you can see that the `OPTIONS` request seems to start _after_ the `GET` request finishes.

<img width="952" alt="screen shot 2015-08-16 at 13 52 20" src="https://cloud.githubusercontent.com/assets/761444/9294539/36540054-441e-11e5-98fd-a754c19b363b.png">

**Comment by timwis** [comment]
Strike that, I've seen it happen in development too

**Comment by timwis** [comment]
So I added a `console.log()` [just before](https://github.com/JarvusInnovations/jarvus-postgrest-apikit/blob/master/src/model/Postgrest.js#L43) and [just after](https://github.com/JarvusInnovations/jarvus-postgrest-apikit/blob/master/src/model/Postgrest.js#L94) the `OPTIONS` request, as well as [just before](https://github.com/timwis/ext-postgrest-crud/blob/master/app/view/main/List.js#L43)/after my store load. In development, the former always begins & ends before the latter begins. In production, both begin around the same time and the store ends first. See console log with timestamps below.

<img width="387" alt="screen shot 2015-08-21 at 06 05 29" src="https://cloud.githubusercontent.com/assets/761444/9406047/06d0e602-47cb-11e5-8ce8-5d3deab9a8da.png">

I'm thinking of firing an event on the model when the `OPTIONS` request completes and listening to that event. Let me know if this is off-base.

**Comment by timwis** [comment]
Could this be the problem? From [Ext.Class source](https://docs.sencha.com/extjs/6.0/6.0.0-classic/source/Class.html#Ext-Class): `Note that the creation process can be asynchronous depending on the pre-processors used.`

``` javascript
/**
     * @method constructor
     * Create a new anonymous class.
     *
     * @param {Object} data An object represent the properties of this class
     * @param {Function} onCreated Optional, the callback function to be executed when this class is fully created.
     * Note that the creation process can be asynchronous depending on the pre-processors used.
     *
     * @return {Ext.Base} The newly created class
     */
    Ext.Class = ExtClass = function(Class, data, onCreated) {
```

**Comment by themightychris** [comment]
Preprocessors are supposed to be allowed to be async. I haven't seen this
issue under 5.1.1. I wonder if something is broken in 6's class loader? I
reviewed the relevant framework code and it doesn't appear to have changed,
and other async class loading seems to still work correctly

On Fri, Aug 21, 2015 at 6:17 AM, Tim Wisniewski notifications@github.com
wrote:

> Could this be the problem? From Ext.Class source
> https://docs.sencha.com/extjs/6.0/6.0.0-classic/source/Class.html#Ext-Class:
> Note that the creation process can be asynchronous depending on the
> pre-processors used.
> 
> /*\*     \* @method constructor     \* Create a new anonymous class.     \*     \* @param {Object} data An object represent the properties of this class     \* @param {Function} onCreated Optional, the callback function to be executed when this class is fully created.     \* Note that the creation process can be asynchronous depending on the pre-processors used.     \*     \* @return {Ext.Base} The newly created class     */
>     Ext.Class = ExtClass = function(Class, data, onCreated) {
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/JarvusInnovations/jarvus-postgrest-apikit/issues/5#issuecomment-133364412
> .

## 

Chris Alfano
Chief Developer
Jarv.us Innovations http://jarv.us

(e) chris@jarv.us (m) 267-918-6166

**Comment by timwis** [comment]
Posted in their [forum](https://www.sencha.com/forum/showthread.php?304610-onBeforeCreated-with-Ajax-request-should-be-synchronous-but-it-s-not-behaving-that-wa&p=1113106#post1113106)
